gstd/gstd_action.c

```diff
@@ -0,0 +1,240 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "gstd_action.h"
+
+enum
+{
+  PROP_TARGET = 1,
+  N_PROPERTIES
+};
+
+#define DEFAULT_PROP_TARGET NULL
+
+/* Gstd Action debugging category */
+GST_DEBUG_CATEGORY_STATIC (gstd_action_debug);
+#define GST_CAT_DEFAULT gstd_action_debug
+
+#define GSTD_DEBUG_DEFAULT_LEVEL GST_LEVEL_INFO
+
+/**
+ * GstdAction:
+ * A wrapper for the conventional action
+ */
+
+G_DEFINE_TYPE (GstdAction, gstd_action, GSTD_TYPE_OBJECT);
+
+/* VTable */
+static void gstd_action_get_property (GObject *, guint, GValue *, GParamSpec *);
+static void gstd_action_set_property (GObject *, guint, const GValue *,
+    GParamSpec *);
+static void gstd_action_dispose (GObject *);
+static GstdReturnCode gstd_action_to_string (GstdObject * obj,
+    gchar ** outstring);
+GstdReturnCode gstd_action_create_default (GstdObject * object,
+    const gchar * name, const gchar * description);
+
+static void
+gstd_action_class_init (GstdActionClass * klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+  GstdObjectClass *gstdc = GSTD_OBJECT_CLASS (klass);
+  GParamSpec *properties[N_PROPERTIES] = { NULL, };
+  guint debug_color;
+
+  object_class->set_property = gstd_action_set_property;
+  object_class->get_property = gstd_action_get_property;
+  object_class->dispose = gstd_action_dispose;
+
+  properties[PROP_TARGET] =
+      g_param_spec_object ("target",
+      "Target",
+      "The target object owning the action",
+      G_TYPE_OBJECT,
+      G_PARAM_READWRITE |
+      G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS | GSTD_PARAM_READ);
+
+
+  g_object_class_install_properties (object_class, N_PROPERTIES, properties);
+
+  gstdc->to_string = GST_DEBUG_FUNCPTR (gstd_action_to_string);
+  gstdc->create = GST_DEBUG_FUNCPTR (gstd_action_create_default);
+
+  /* Initialize debug category with nice colors */
+  debug_color = GST_DEBUG_FG_BLACK | GST_DEBUG_BOLD | GST_DEBUG_BG_WHITE;
+  GST_DEBUG_CATEGORY_INIT (gstd_action_debug, "gstdaction", debug_color,
+      "Gstd Action category");
+
+}
+
+static void
+gstd_action_init (GstdAction * self)
+{
+  self->target = DEFAULT_PROP_TARGET;
+}
+
+static void
+gstd_action_dispose (GObject * object)
+{
+  GstdAction *self = GSTD_ACTION (object);
+
+  if (self->target) {
+    g_object_unref (self->target);
+    self->target = NULL;
+  }
+
+  G_OBJECT_CLASS (gstd_action_parent_class)->dispose (object);
+}
+
+static void
+gstd_action_get_property (GObject * object,
+    guint property_id, GValue * value, GParamSpec * pspec)
+{
+  GstdAction *self = GSTD_ACTION (object);
+
+  switch (property_id) {
+    case PROP_TARGET:
+      GST_DEBUG_OBJECT (self, "Returning action owner %p (%s)", self->target,
+          GST_OBJECT_NAME (self->target));
+      g_value_set_object (value, self->target);
```

Comment:
Since properties are accessed from another thread, we should serialize access to the `self->target` with `GST_OBJECT_LOCK`. (I know other objects lack this, but it's something we need to fix).